File: android/robot/src/main/java/org/openbot/env/ConnectionSelector.java
package org.openbot.env;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.wifi.WifiManager;

import java.lang.reflect.Method;

public class ConnectionSelector {
  private static final String TAG = "ConnectionManager";
  private static ConnectionSelector _connectionSelector;
  private static Context _context;
  private ILocalConnection connection;

  private final ILocalConnection networkConnection = new NetworkServiceConnection();
  private final ILocalConnection nearbyConnection = new NearbyConnection();

  private ConnectionSelector() {
    if (_connectionSelector != null) {
      throw new RuntimeException(
          "Use getInstance() method to get the single instance of this class.");
    }
  }

  public static ConnectionSelector getInstance(Context context) {
    _context = context;
    if (_connectionSelector == null) {

      synchronized (ConnectionSelector.class) {
        if (_connectionSelector == null) _connectionSelector = new ConnectionSelector();
      }
    }

    return _connectionSelector;
  }

  ILocalConnection getConnection() {
    if (connection != null) {
      return connection;
    }

    if (isConnectedViaWifi() || isWifiApEnabled()) {
      connection = networkConnection;
    } else {
      connection = nearbyConnection;
    }

    return connection;
  }

  private boolean isConnectedViaWifi() {
    ConnectivityManager connectivityManager =
        (ConnectivityManager) _context.getSystemService(Context.CONNECTIVITY_SERVICE);
    NetworkInfo mWifi = connectivityManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
    return mWifi.isConnected();
  }

  private boolean isWifiApEnabled() {
    WifiManager wifiManager =
            (WifiManager) _context.getSystemService(Context.WIFI_SERVICE);
    try {
      final Method method =
              wifiManager.getClass().getDeclaredMethod("isWifiApEnabled");
      Boolean result = (Boolean) method.invoke(wifiManager);
      return Boolean.TRUE.equals(result);
    } catch (final Throwable ignored) {}

    return false;
  }
}

File: android/controller/src/main/java/org/openbot/controller/ConnectionSelector.kt
package org.openbot.controller

import android.annotation.SuppressLint
import android.content.Context
import android.content.Context.WIFI_SERVICE
import android.net.wifi.WifiManager

@SuppressLint("StaticFieldLeak")
object ConnectionSelector {

    private val connection: ILocalConnection? = null
    private const val TAG = "ConnectionManager"
    @SuppressLint("StaticFieldLeak")
    private lateinit var context: Context

    fun init (context: Context) {
        this.context = context
    }

    fun getConnection(): ILocalConnection {
        val connected = isConnectedViaWifi(context) || isWifiApEnabled(context)
        return if (connected) NetworkServiceConnection else NearbyConnection
    }

    private fun isConnectedViaWifi(context: Context): Boolean {

        val wifiManager = context.getSystemService(WIFI_SERVICE) as WifiManager?
        val info = wifiManager!!.connectionInfo ?: return false
        val networkId = info.networkId

        return networkId > 0
    }

    private fun isWifiApEnabled(context: Context): Boolean {
        val wifiManager = context.getSystemService(WIFI_SERVICE) as WifiManager
        return try {
            val method = wifiManager.javaClass.getDeclaredMethod("isWifiApEnabled")
            method.invoke(wifiManager) as? Boolean == true
        } catch (ignored: Throwable) {
            false
        }
    }
}
